### ISSUE TITLE
Possibility to resume from specific binlog and position

### ISSUE THREAD

**Issue by DevMasterSoonToBe** [cashapp/spirit#365]
Hi, 

Is it possible to give Spirit binlog file and position from which to resume altering?

Our use case:

We have an issue that binlog accidentaly purged on our instance and spirit could not resume from needed binlog

INFO[27855] closing canal
INFO[27855] syncer is closing...
WARN[27855] releasing metadata lock for x.xx
ERRO[27855] error writing checkpoint: context canceled
INFO[27855] kill last connection id xxx
INFO[27855] syncer is closed
spirit: error: Error 1205 (HY000): Lock wait timeout exceeded; try restarting transaction
[root@ip-xxxxx spirit]# ./spirit --host=${DB_HOST} --username=${DB_USERNAME} --password=${DB_PASSWORD} --database=x --table=xx --alter="PARTITION BY KEY() PARTITIONS 8" --defer-cutover=true --checksum=true --skip-drop-after-cutover=true --threads=32 --target-chunk-time=1s
INFO[0000] Starting spirit migration: concurrency=32 target-chunk-size=1s table=x.xx alter="PARTITION BY KEY() PARTITIONS 8"
INFO[0000] attempting to acquire metadata lock for x.xx
INFO[0000] acquired metadata lock: x.xx
INFO[0000] unable to use INPLACE: ALTER either does not support INPLACE or when performed as INPLACE could take considerable time. Use --force-inplace to override this safety check
INFO[0006] create BinlogSyncer with config {ServerID:1730 Flavor:mysql Host:xxxxxxxxxxx Port:3306 User:xxxxxx Password: Localhost: Charset:utf8 SemiSyncEnabled:false RawModeEnabled:false TLSConfig:0xc000504540 ParseTime:false TimestampStringLocation:UTC UseDecimal:false RecvBufferSize:0 HeartbeatPeriod:0s ReadTimeout:0s MaxReconnectAttempts:0 DisableRetrySync:false VerifyChecksum:false DumpCommandFlag:0 Option:<nil> Logger:0xc000046970 Dialer:0x7c2600 RowsEventDecodeFunc:0xa95a20 TableMapOptionalMetaDecodeFunc:<nil> DiscardGTIDSet:false EventCacheCount:10240}
WARN[0007] resuming from checkpoint failed because resuming from the previous binlog position failed. log-file: mysql-bin-changelog.x log-pos: xxxxx
INFO[0007] could not resume from checkpoint: reason=binlog position is impossible, the source may have already purged it

We have restored instance from snapshot and possibly want to resume spirit by giving specific binlog and position to start from. Is that possible?

**Comment by DevMasterSoonToBe** [comment]
@morgo  Perhaps wider question - Can we resume spirit on restored DB with chkpnt, sentinel, _new tables there and it won't have any data incosistencies between source and new tables?

**Comment by DevMasterSoonToBe** [comment]
Also, 

1. can spirit run on read replica? We have B/G deployment and on green environment it runs ok, wondering if on Read Replica it will run as well.
2. ERRO[27855] error writing checkpoint: context canceled what does this error mean behind the scenes in Spirit? Does it mean binlog is not getting updated to newest checkpoint? Because we have binlog retention period set to 24hrs and even though 24hrs have not passed spirit tried to resume from binlog older than 24hrs.

**Comment by morgo** [comment]
> [@morgo](https://github.com/morgo) Perhaps wider question - Can we resume spirit on restored DB with chkpnt, sentinel, _new tables there and it won't have any data incosistencies between source and new tables?

It may have inconsistencies if it misses some binary logs. This will happen if a modification occurs below the current checkpoint, because a change can occur which is not identified.

However, having said that, the lost change can be identified by the checkpoint, and the checkpoint will repair inconsistencies as they are discovered.

I don't think this is something we will ever support as an official feature, but you can manually edit the checkpoint table to set the binary log value to a current existing position in time. When you resume, you *must* have the checksum enabled to prevent data loss.

> can spirit run on read replica? We have B/G deployment and on green environment it runs ok, wondering if on Read Replica it will run as well.

This is not something we have planned. It is unlikely to benefit Spirit that much, since it uses `INSERT IGNORE .. SELECT FROM` and doesn't actually read the data into Spirit.

> ERRO[27855] error writing checkpoint: context canceled what does this error mean behind the scenes in Spirit? Does it mean binlog is not getting updated to newest checkpoint? Because we have binlog retention period set to 24hrs and even though 24hrs have not passed spirit tried to resume from binlog older than 24hrs.

You are hitting this issue: https://github.com/cashapp/spirit/issues/197

Basically when a checkpoint can't be written (perhaps because the system is overloaded) the migration is allowed to continue. But there's a risk with this, since if there's a failure the tracked position in the binary log (as per the checkpoint table) is way behind where the actual applied position of the binary log is. In future, it should be fixed so that failing to write a checkpoint causes a failure for spirit.

**Comment by DevMasterSoonToBe** [comment]
**It may have inconsistencies if it misses some binary logs. This will happen if a modification occurs below the current checkpoint, because a change can occur which is not identified.

However, having said that, the lost change can be identified by the checksum, and the checksum will repair inconsistencies as they are discovered.

I don't think this is something we will ever support as an official feature, but you can manually edit the checkpoint table to set the binary log value to a current existing position in time. When you resume, you must have the checksum enabled to prevent data loss.**

Regarding above - in theory we have binlog file from which we should resume in the output below, right? So perhaps as you suggest editing checkpoint table to set it to that value/file.

WARN[0007] resuming from checkpoint failed because resuming from the previous binlog position failed. log-file: mysql-bin-changelog.x log-pos: xxxxx


**This is not something we have planned. It is unlikely to benefit Spirit that much, since it uses INSERT IGNORE .. SELECT FROM and doesn't actually read the data into Spirit.**

Could you elaborate on this? Technically green environment instances are also replicas from blue environment, so I would assume if it works on B/G setup it should work correctly on Read Replica which would later could be promoted to master.


**You are hitting this issue: https://github.com/cashapp/spirit/issues/197

Basically when a checkpoint can't be written (perhaps because the system is overloaded) the migration is allowed to continue. But there's a risk with this, since if there's a failure the tracked position in the binary log (as per the checkpoint table) is way behind where the actual applied position of the binary log is. In future, it should be fixed so that failing to write a checkpoint causes a failure for spirit.**

Thank you, it makes more sense now. And there is no way to know what is the actual applied position of the binary log? What happens if in checkpoint table I set it to binlog which is way begind the actual applied position? And can there be any other reasons besides system overload? From metrics I don't see any bottlenecks.

**Comment by DevMasterSoonToBe** [comment]
Perhaps on the last point, can replica lag be a factor? This is only metric I am seeing that is having issues(more than 10k seconds lag).

Also, what is best/recommended binlog_retention_period to be set on DB where spirit runs? If it has issues like you described with writing to checkpoint table I reckon bigger retention period could help preventing this from user side.

**Comment by morgo** [comment]
> I don't think this is something we will ever support as an official feature, but you can manually edit the checkpoint table to set the binary log value to a current existing position in time. When you resume, you must have the checksum enabled to prevent data loss.**
> 
> Regarding above - in theory we have binlog file from which we should resume in the output below, right? So perhaps as you suggest editing checkpoint table to set it to that value/file.

Yes, provided these points are still available.
 
> WARN[0007] resuming from checkpoint failed because resuming from the previous binlog position failed. log-file: mysql-bin-changelog.x log-pos: xxxxx
> 
> **This is not something we have planned. It is unlikely to benefit Spirit that much, since it uses INSERT IGNORE .. SELECT FROM and doesn't actually read the data into Spirit.**
> 
> Could you elaborate on this? Technically green environment instances are also replicas from blue environment, so I would assume if it works on B/G setup it should work correctly on Read Replica which would later could be promoted to master.

If you are using B/G or your read-replicas are writable, you may be able apply the schema changes directly. If it's a blocking change, replication will take some time to catch up after its complete. I have seen this method suggested in the AWS documentation. I don't recommend it though; what they are not telling you, is that Row-based replication depends strictly on the ordinality of columns. If you were to perform a schema change where you remove a column mid-way through, and add a new column to the end of the table, you could end up with data corruption as changes are incorrectly applied on the replica.

Gh-ost for example supports read-replicas for estimates (number of rows in the table, where chunk boundaries should start/stop). We use a different method for table estimates in Spirit which is not quite as accurate, and for chunk boundaries we have 2 different implementations of chunker. But if we need to read-and-estimate I actually think it is beneficial to do it on the writer because it will pre-load those pages into memory, ahead of them being copied shortly after.

> **You are hitting this issue: [#197](https://github.com/cashapp/spirit/issues/197)
> 
> Basically when a checkpoint can't be written (perhaps because the system is overloaded) the migration is allowed to continue. But there's a risk with this, since if there's a failure the tracked position in the binary log (as per the checkpoint table) is way behind where the actual applied position of the binary log is. In future, it should be fixed so that failing to write a checkpoint causes a failure for spirit.**
> 
> Thank you, it makes more sense now. And there is no way to know what is the actual applied position of the binary log? What happens if in checkpoint table I set it to binlog which is way begind the actual applied position? And can there be any other reasons besides system overload? From metrics I don't see any bottlenecks.

It may also be that there are some connections in the connection pool which need to be cycled out. When we hit this last, we hadn't added `rejectReadOnly` to the DSN yet, but it should be there now. We haven't hit this issue in a while, so I can't say with certainty what causes it. But if you have a reproducible testcase that shows it's quite probably we will be happy to prioritize a fix.

**Comment by DevMasterSoonToBe** [comment]
@morgo Thanks!

One last thing before we close this case:

INFO[14173] syncer is closing...
WARN[14173] releasing metadata lock for x.xx
ERRO[14173] error flushing binary log: context canceled

INFO[27855] closing canal INFO[27855] syncer is closing... WARN[27855] releasing metadata lock for x.xx ERRO[27855] error writing checkpoint: context canceled INFO[27855] kill last connection id 3479009 INFO[27855] syncer is closed spirit: error: Error 1205 (HY000): Lock wait timeout exceeded; try restarting transaction

Is it logical to think that some other connection(perhaps replication from master) tries to lock same table which we try to partition and spirit prematurely releases metadata lock leading to error which I see in the output above?

I am trying to think of root causes why randomly spirit just dies sometimes for us.

**Comment by morgo** [comment]
> INFO[27855] closing canal INFO[27855] syncer is closing... WARN[27855] releasing metadata lock for x.xx ERRO[27855] error writing checkpoint: context canceled INFO[27855] kill last connection id 3479009 INFO[27855] syncer is closed spirit: error: Error 1205 (HY000): Lock wait timeout exceeded; try restarting transaction
> 
> Is it logical to think that some other connection(perhaps replication from master) tries to lock same table which we try to partition and spirit prematurely releases metadata lock leading to error which I see in the output above?
> 
> I am trying to think of root causes why randomly spirit just dies sometimes for us.

The metadata lock is a `get_lock()` call. It's an advisory lock, not an actual lock, so it will only block other spirit processes from running and not user queries.

I think I have a better answer for you! The most pertinent error here is this one:
> Error 1205 (HY000): Lock wait timeout exceeded; try restarting transaction

This could either be an InnoDB row-lock (trying to copy a chunk) or a metadata lock (trying to acquire a lock for checksum or cutover).

For InnoDB row-locks, the MySQL default is 50s, but spirit lowers this to 3s here: https://github.com/cashapp/spirit/blob/a04a2eeb48d03f42b3d8c38117e45d3cfe83c8da/pkg/dbconn/dbconn.go#L39

(We don't want spirit to be blamed for impacting production systems, so we are unlikely to make this configurable.)

For metadata lock on cutover, the MySQL default is 8hours, but the spirit default is 30s. We do make this one configurable, but we don't document it. Try `--lock-wait-timeout=120s` and let me know how it goes!

(Caveat being that this could bring your system offline for 2 minutes if it can't acquire that lock. Again, we default to a lower timeout so spirit doesn't get blamed for taking systems offline.)

**Comment by DevMasterSoonToBe** [comment]
@morgo do you recommend changing this on DB(it has param like this) or spirit has this flag? Because in Usage.md I don't see this described

**Comment by morgo** [comment]
> [@morgo](https://github.com/morgo) do you recommend changing this on DB(it has param like this) or spirit has this flag? Because in Usage.md I don't see this described

This is a spirit flag - sorry it's not documented, we can add it.

Changing it on the DB (such as with a parameter group) will have no effect because spirit always overwrites it.

**Comment by DevMasterSoonToBe** [comment]
But basically what I can gather from this - **this could either be an InnoDB row-lock (trying to copy a chunk) or a metadata lock (trying to acquire a lock for checksum or cutover).**

It is still for us in process of copying data(not doing checksum or cutover), so possibly **this could either be an InnoDB row-lock (trying to copy a chunk)** is probable here.

But how does it work? Does it mean spirit itself kind of is the cause of failure?

**Comment by DevMasterSoonToBe** [comment]
FYI, we are trying to partition very large table(more than 15TB). and target-chunk-time is set to 1s(due to high waiting time for rollback transactions to finish when spirit fails).

**Comment by morgo** [comment]
> It is still for us in process of copying data(not doing checksum or cutover), so possibly **this could either be an InnoDB row-lock (trying to copy a chunk)** is probable here.
> 
> But how does it work? Does it mean spirit itself kind of is the cause of failure?

No, this likely has more to do with your application.

Before you modify a row, locks have to be acquired. Those locks are then held for the duration of the transaction (not the statement!), and are only released when you commit or rollback.

Each chunk-copy is it's own transaction, so Spirit is ostensibly performing 1s transactions in your case. Spirit could technically conflict with itself (between applying rows and copier), but the copier won't conflict with itself. The more likely scenario is it conflicts with your application, and the more likely scenario of the two is that the replication applier causes the conflicts, because you tend to modify rows that have recently been modified.

When a row-lock can't be acquired immediately it will queue waiting, and then other row locks will queue waiting behind it. Chunking (at 1s) is for two reasons: replication lag, and yielding locks so that foreground application threads will not have timeouts triggered by spirit. So you can see why the `innodb-lock-wait-timeout` of 3s is also closely related to this goal. Spirit is likely timing out because an application thread is locking and processing these rows.

You are not seeing these problems in your application, because the default MySQL timeouts are crazy high (50s for innodb locks, and 8 hours for data locks). For many OLTP outages, waiting until these timeouts will basically cause an outage.

You can see running transactions with:
```
select * from information_schema.INNODB_TRX\G
```

And locks those transactions hold with:
```
select * from performance_schema.data_locks\G
```

Be careful that the `data_locks` query is pretty expensive prior to 8.0.40.

It might be that your transactions are too long-lived. If you can identify the cause (for example: maybe you are holding a transaction open while you do another network request?), it's not just a fix for spirit, but your application will perform better.

**Comment by morgo** [comment]
I closed this because we fixed the missing docs case, but otherwise believe spirit is working as intended. Please feel free to re-open if you continue to experience issues. Thanks!
